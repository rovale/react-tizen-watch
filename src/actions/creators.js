import * as actionType from './types';

export const initializeApp = () => ({ type: actionType.INITIALIZE_APP });
export const handleError = (msg, url, lineNo, columnNo, error) => ({ type: actionType.HANDLE_ERROR, payload: { msg, url, lineNo, columnNo, error } });
export const loadPages = pages => ({ type: actionType.LOAD_PAGES, payload: { pages } });
export const loadDevices = devices => ({ type: actionType.LOAD_DEVICES, payload: { devices } });

export const activatePage = id => ({ type: actionType.ACTIVATE_PAGE, payload: { id } });
export const selectPage = id => ({ type: actionType.SELECT_PAGE, payload: { id } });
export const closePage = () => ({ type: actionType.CLOSE_PAGE });

export const activateDevice = id => ({ type: actionType.ACTIVATE_DEVICE, payload: { id } });
export const selectDevice = id => ({ type: actionType.SELECT_DEVICE, payload: { id } });
export const closeDevice = () => ({ type: actionType.CLOSE_DEVICE });
export const changeDeviceAttribute = event => ({ type: actionType.CHANGE_DEVICE_ATTRIBUTE, payload: { event } });

export const toggleSwitch = id => ({ type: actionType.TOGGLE_SWITCH, payload: { id } });

export const activateButton = id => ({ type: actionType.ACTIVATE_BUTTON, payload: { id } });
export const clickButton = id => ({ type: actionType.CLICK_BUTTON, payload: { id } });
