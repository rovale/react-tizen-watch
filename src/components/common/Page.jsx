import React, { useRef } from 'react';
import PropTypes from 'prop-types';
import scroll from 'smooth-move';

export const ScrollContext = React.createContext();

export const Page = ({ children }) => {
  const element = useRef(null);

  const scrollToMiddle = (duration, itemTop, itemHeight) => {
    const scroller = element.current.getElementsByClassName('ui-scroller')[0];
    const scrollerHeight = scroller.getBoundingClientRect().height;
    const newScrollTop = (itemTop + scroller.scrollTop) - ((scrollerHeight - itemHeight) / 2);

    scroll(scroller, { x: 0, y: newScrollTop, duration });
  };

  return (
    <div ref={element} className="ui-page ui-page-active ui-scroll-on">
      <div className="ui-scroller ui-snap-container" ref={e => e && e.setAttribute('tizen-circular-scrollbar', '')}>
        <ScrollContext.Provider value={scrollToMiddle}>
          {children}
        </ScrollContext.Provider>
      </div>
    </div>
  );
};

Page.propTypes = {
  children: PropTypes.node.isRequired,
};

export const Header = ({ children }) => (
  <header className="ui-header">
    <h2 className="ui-title">{children}</h2>
  </header>
);

Header.propTypes = {
  children: PropTypes.string.isRequired,
};

export const Content = ({ children }) => {
  return (
    <div className="ui-content" ref={e => e && e.setAttribute('tizen-circular-scrollbar', '')}>
      {children}
    </div>);
};

Content.propTypes = {
  children: PropTypes.node.isRequired,
};
