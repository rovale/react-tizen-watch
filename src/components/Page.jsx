import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import * as action from '../actions/creators';
import { Page as UiPage, Header, Content } from './common/Page';
import { List } from './common/List';
import Switch from './deviceItem/Switch';
import Display from './deviceItem/Display';
import Presence from './deviceItem/Presence';
import Temperature from './deviceItem/Temperature';
import Thermostat from './deviceItem/Thermostat';

let Page = ({
  pageTitle, devices, activeDeviceId, onActivateDevice, onSelectDevice, onToggleSwitch,
}) => {
  const getListItemContent = (device) => {
    switch (device.template) {
      case 'device':
        return <Display device={device} />;
      case 'presence':
        return <Presence device={device} />;
      case 'switch':
        return <Switch device={device} />;
      case 'temperature':
        return <Temperature device={device} />;
      case 'thermostat':
        return <Thermostat device={device} />;
      default:
        return <div>{`${device.title} - (${device.template})`}</div>;
    }
  };

  const getOnSelect = (device) => {
    switch (device.template) {
      case 'switch':
        return () => onToggleSwitch(device.id);
      default:
        return null;
    }
  };

  const deviceOptions = devices.map(d => ({
    ...d,
    content: getListItemContent(d),
    onSelect: getOnSelect(d),
  }));

  return (
    <UiPage>
      <Header>{pageTitle}</Header>
      <Content>
        <List
          options={deviceOptions}
          activeOptionId={activeDeviceId}
          onActivateOption={onActivateDevice}
          onSelectOption={onSelectDevice}
        />
      </Content>
    </UiPage>);
};

Page.propTypes = {
  pageTitle: PropTypes.string.isRequired,
  devices: PropTypes.arrayOf(PropTypes.object).isRequired,
  activeDeviceId: PropTypes.string.isRequired,
  onActivateDevice: PropTypes.func.isRequired,
  onSelectDevice: PropTypes.func.isRequired,
  onToggleSwitch: PropTypes.func.isRequired,
};

const mapStateToProps = (state) => {
  const selectedPage = state.pages.find(p => p.id === state.ui.selectedPageId);

  return {
    pageTitle: selectedPage.title,
    devices: selectedPage.devices.map(pd => state.devices.find(d => pd === d.id)),
    activeDeviceId: state.ui.activeDeviceId || selectedPage.devices[0],
  };
};

const mapDispatchToProps = dispatch => ({
  onActivateDevice: id => dispatch(action.activateDevice(id)),
  onSelectDevice: id => dispatch(action.selectDevice(id)),
  onToggleSwitch: id => dispatch(action.toggleSwitch(id)),
});

export default Page = connect(mapStateToProps, mapDispatchToProps)(Page);
