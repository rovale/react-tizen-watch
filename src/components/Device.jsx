import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import { Page as UiPage, Header, Content } from './common/Page';
import ButtonsDevice from './ButtonsDevice';
import * as action from '../actions/creators';

let Device = ({ device, onClose }) => {
  const getDeviceContent = () => {
    switch (device.template) {
      case 'buttons':
        return <ButtonsDevice device={device} />;
      default:
        return <div>Content of {device.id} {device.template}</div>;
    }
  };

  return (<UiPage>
    <Header>{device.title}</Header>
    <Content>
      {getDeviceContent(device)}
    </Content>
    {/* <footer className="ui-footer ui-bottom-button ui-fixed">
      <a href="#close" onClick={onClose} className="ui-btn">OK</a>
    </footer> */}
  </UiPage>);
};

Device.propTypes = {
  device: PropTypes.object.isRequired,
  onClose: PropTypes.func.isRequired,
};

const mapStateToProps = state => ({
  device: state.devices.find(d => d.id === state.ui.selectedDeviceId),
});

const mapDispatchToProps = dispatch => ({
  onClose: () => dispatch(action.closeDevice()),
});

Device = connect(mapStateToProps, mapDispatchToProps)(Device);

export default Device;
