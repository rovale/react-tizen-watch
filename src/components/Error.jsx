import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import { Page, Header, Content } from './common/Page';

const Error = ({
  msg, url, lineNo, columnNo, error,
}) => (
  <Page>
    <Header>Error</Header>
    <Content>
      Msg: {msg}<br />
      Url: {url}<br />
      LineNo: {lineNo}<br />
      ColumnNo: {columnNo}<br />
      Error: {!!error && error.stack }
    </Content>
  </Page>
);

Error.propTypes = {
  msg: PropTypes.string.isRequired,
  url: PropTypes.string.isRequired,
  lineNo: PropTypes.number.isRequired,
  columnNo: PropTypes.number.isRequired,
  error: PropTypes.shape({
    stack: PropTypes.string.isRequired,
  }).isRequired,
};

const mapStateToProps = state => ({
  msg: state.ui.error.msg,
  url: state.ui.error.url,
  lineNo: state.ui.error.lineNo,
  columnNo: state.ui.error.columnNo,
  error: state.ui.error.error,
});

export default connect(mapStateToProps)(Error);
