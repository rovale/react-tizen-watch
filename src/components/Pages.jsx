import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import * as action from '../actions/creators';
import { Page as UiPage, Header, Content } from './common/Page';
import { List } from './common/List';

let Pages = ({
  pages, activePageId, onActivatePage, onSelectPage,
}) => (
  <UiPage>
    <Header>React Tizen</Header>
    <Content>
      <List
        options={pages}
        activeOptionId={activePageId}
        onActivateOption={onActivatePage}
        onSelectOption={onSelectPage}
      />
    </Content>
  </UiPage>
);

Pages.propTypes = {
  pages: PropTypes.arrayOf(PropTypes.object).isRequired,
  activePageId: PropTypes.string.isRequired,
  onActivatePage: PropTypes.func.isRequired,
  onSelectPage: PropTypes.func.isRequired,
};

const mapStateToProps = state => ({
  pages: state.pages,
  activePageId: state.ui.activePageId || state.pages[0].id,
});

const mapDispatchToProps = dispatch => ({
  onActivatePage: id => dispatch(action.activatePage(id)),
  onSelectPage: id => dispatch(action.selectPage(id)),
});

Pages = connect(mapStateToProps, mapDispatchToProps)(Pages);

export default Pages;
