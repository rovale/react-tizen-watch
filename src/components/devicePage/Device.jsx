import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import { Page as UiPage, Header, Content } from '../common/Page';
import ButtonsDevice from './ButtonsDevice';

const Device = ({ device }) => {
  const getDeviceContent = () => {
    switch (device.template) {
      case 'buttons':
        return <ButtonsDevice device={device} />;
      default:
        return <div>Content of {device.id} {device.template}</div>;
    }
  };

  return (
    <UiPage>
      <Header>{device.title}</Header>
      <Content>
        {getDeviceContent(device)}
      </Content>
    </UiPage>
  );
};

Device.propTypes = {
  device: PropTypes.object.isRequired,
};

const mapStateToProps = state => ({
  device: state.devices.find(d => d.id === state.ui.selectedDeviceId),
});

export default connect(mapStateToProps)(Device);
