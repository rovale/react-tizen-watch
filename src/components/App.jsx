import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import * as action from '../actions/creators';
import Error from './Error';
import Splash from './Splash';
import Pages from './Pages';
import Page from './Page';
import Device from './devicePage/Device';

class App extends Component {
  static propTypes = {
    mainRoute: PropTypes.string,
    selectedPageId: PropTypes.string,
    selectedDeviceId: PropTypes.string,
    onClosePage: PropTypes.func.isRequired,
    onCloseDevice: PropTypes.func.isRequired,
  };

  static defaultProps = {
    mainRoute: null,
    selectedPageId: null,
    selectedDeviceId: null,
  }

  componentDidMount() {
    window.addEventListener('tizenhwkey', (ev) => {
      if ((ev.key || ev.keyName) === 'back') {
        if (!this.props.selectedPageId) {
          try {
            window.tizen.application.getCurrentApplication().exit();
          } catch (err) {
            // ignore
          }
        } else if (this.props.selectedPageId && !this.props.selectedDeviceId) {
          this.props.onClosePage();
        } else {
          this.props.onCloseDevice();
        }
      }
    });

    window.addEventListener('keyup', (e) => {
      if (e.key === 'ArrowRight') {
        window.dispatchEvent(new CustomEvent('rotarydetent', { detail: { direction: 'CW' } }));
      } else if (e.key === 'ArrowLeft') {
        window.dispatchEvent(new CustomEvent('rotarydetent', { detail: { direction: 'CCW' } }));
      } else if (e.key === 'Backspace') {
        window.dispatchEvent(new KeyboardEvent('tizenhwkey', { key: 'back' }));
      }
    });
  }

  render() {
    return (
      <div>
        {this.props.mainRoute === 'error' && <Error />}
        {this.props.mainRoute === 'splash' && <Splash />}
        {this.props.mainRoute === 'pages' && !this.props.selectedPageId && <Pages />}
        {this.props.mainRoute === 'pages' && this.props.selectedPageId && !this.props.selectedDeviceId && <Page />}
        {this.props.mainRoute === 'pages' && this.props.selectedPageId && this.props.selectedDeviceId && <Device />}
      </div>
    );
  }
}

const mapStateToProps = state => ({
  mainRoute: state.ui.mainRoute,
  selectedPageId: state.ui.selectedPageId,
  selectedDeviceId: state.ui.selectedDeviceId,
});

const mapDispatchToProps = dispatch => ({
  onClosePage: () => dispatch(action.closePage()),
  onCloseDevice: () => dispatch(action.closeDevice()),
});

export default connect(mapStateToProps, mapDispatchToProps)(App);
