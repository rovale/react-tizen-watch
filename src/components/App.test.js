import React from 'react';
import Enzyme, { shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import { createStore } from 'redux';
import reducer from '../reducers';
import * as action from '../actions/creators';

import Error from './Error';
import Splash from './Splash';
import Pages from './Pages';
import Page from './Page';
import Device from './devicePage/Device';
import App from './App';

Enzyme.configure({ adapter: new Adapter() });

describe('The App component', () => {
  let store;

  beforeEach(() => {
    store = createStore(reducer);
  });

  test('it should render the Splash component', () => {
    store.dispatch(action.initializeApp());

    const app = mount((<App store={store} />));

    expect(app.find(Splash).exists()).toBe(true);
    expect(app.find(Splash)).toMatchSnapshot();
  });

  test('it should render the Error component', () => {
    store.dispatch(action.handleError('someMsg', 'someUrl', 10, 20, { stack: 'someErrorStack' }));

    const app = shallow((<App store={store} />));

    expect(app.dive().find(Error).exists()).toBe(true);
  });

  test('it should render the Pages component', () => {
    store.dispatch(action.loadPages([]));

    const app = shallow((<App store={store} />));

    expect(app.dive().find(Pages).exists()).toBe(true);
  });

  test('it should render the Page component', () => {
    const pages = [{ id: 'climate', devices: [] }];
    store.dispatch(action.loadPages(pages));
    store.dispatch(action.selectPage('climate'));

    const app = shallow((<App store={store} />));

    expect(app.dive().find(Page).exists()).toBe(true);
  });

  test('it should render the Device component', () => {
    const pages = [{ id: 'climate', devices: ['thermostat'] }];
    const devices = [{ id: 'thermostat' }];
    store.dispatch(action.loadPages(pages));
    store.dispatch(action.loadDevices(devices));
    store.dispatch(action.selectPage('climate'));
    store.dispatch(action.selectDevice('thermostat'));

    const app = shallow((<App store={store} />));

    expect(app.dive().find(Device).exists()).toBe(true);
  });

  describe('when the back button is pressed', () => {
    let events;

    beforeEach(() => {
      events = {};
      window.addEventListener = jest.fn().mockImplementation((eventType, cb) => {
        events[eventType] = cb;
      });

      window.dispatchEvent = jest.fn().mockImplementation((event) => {
        events[event.type](event);
      });
    });

    test('it should close the application when showing the splash screen', () => {
      store.dispatch(action.initializeApp());
      shallow((<App store={store} />)).dive();
      const exit = jest.fn();
      window.tizen = { application: { getCurrentApplication: () => ({ exit }) } };

      events.tizenhwkey({ keyName: 'back' });

      expect(exit.mock.calls).toHaveLength(1);
    });

    test('it should close the application when showing the pages', () => {
      store.dispatch(action.loadPages([]));
      shallow((<App store={store} />)).dive();
      const exit = jest.fn();
      window.tizen = { application: { getCurrentApplication: () => ({ exit }) } };

      events.tizenhwkey({ keyName: 'back' });

      expect(exit.mock.calls).toHaveLength(1);
    });

    test('it should close the page when showing a specific page', () => {
      const pages = [{ id: 'climate', devices: [] }];
      store.dispatch(action.loadPages(pages));
      store.dispatch(action.selectPage('climate'));
      shallow((<App store={store} />)).dive();

      events.tizenhwkey({ keyName: 'back' });

      expect(store.getState().ui.selectedPageId).toBeNull();
    });

    test('it should close the device when showing a specific device', () => {
      const pages = [{ id: 'climate', devices: ['thermostat'] }];
      const devices = [{ id: 'thermostat' }];
      store.dispatch(action.loadPages(pages));
      store.dispatch(action.loadDevices(devices));
      store.dispatch(action.selectPage('climate'));
      store.dispatch(action.selectDevice('thermostat'));
      shallow((<App store={store} />)).dive();

      events.tizenhwkey({ keyName: 'back' });

      expect(store.getState().ui.selectedDeviceId).toBeNull();
    });
  });

  describe('when testing the application in a browser', () => {
    let events;

    beforeEach(() => {
      events = {};
      window.addEventListener = jest.fn().mockImplementation((eventType, cb) => {
        events[eventType] = cb;
      });
    });

    test('it should dispatch a back tizenhwkey event on backspace button press', () => {
      const spy = jest.fn();
      window.dispatchEvent = spy;
      store.dispatch(action.loadPages([]));
      shallow((<App store={store} />)).dive();

      events.keyup({ key: 'Backspace' });

      expect(spy.mock.calls).toHaveLength(1);
      expect(spy.mock.calls[0][0].type).toBe('tizenhwkey');
      expect(spy.mock.calls[0][0].key).toBe('back');
    });

    test('it should dispatch a CCW rotarydetent event on left arrow press', () => {
      const spy = jest.fn();
      window.dispatchEvent = spy;
      store.dispatch(action.loadPages([]));
      shallow((<App store={store} />)).dive();

      events.keyup({ key: 'ArrowLeft' });

      expect(spy.mock.calls).toHaveLength(1);
      expect(spy.mock.calls[0][0].type).toBe('rotarydetent');
      expect(spy.mock.calls[0][0].detail).toEqual({ direction: 'CCW' });
    });

    test('it should dispatch a CW rotarydetent event on right arrow press', () => {
      const spy = jest.fn();
      window.dispatchEvent = spy;
      store.dispatch(action.loadPages([]));
      shallow((<App store={store} />)).dive();

      events.keyup({ key: 'ArrowRight' });

      expect(spy.mock.calls).toHaveLength(1);
      expect(spy.mock.calls[0][0].type).toBe('rotarydetent');
      expect(spy.mock.calls[0][0].detail).toEqual({ direction: 'CW' });
    });
  });
});
