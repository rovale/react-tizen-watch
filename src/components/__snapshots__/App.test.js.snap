// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The App component it should render the Splash component 1`] = `
<Splash>
  <Page>
    <div
      className="ui-page ui-page-active ui-scroll-on"
    >
      <div
        className="ui-scroller ui-snap-container"
      >
        <Content>
          <div
            className="ui-content"
          >
            <div
              className="small-processing-container"
            >
              <div
                className="ui-processing"
              />
              <div
                className="ui-processing-text"
              >
                Fetching data...
              </div>
            </div>
          </div>
        </Content>
      </div>
    </div>
  </Page>
</Splash>
`;
